### ISSUE TITLE
New URI pattern policy for identifying DOREMUS resources

### ISSUE THREAD

**Issue by pasqLisena** [DOREMUS-ANR/marc2rdf#21]
Following previous conversation, I try to write out a proposal of pattern for URI.
# Current Pattern

``` turtle
http://data.doremus.org/<class_name>/<class_code>/<uuid>
```

where the UUID is computed following the rules described in #16 .

Example:

``` turtle
http://data.doremus.org/Self_Contained_Expression/F22/10430039-5e02-3ec8-bc6b-b433cdde13c7
```
# Proposal

We think about a simplification of the uri. The new pattern can be:

``` turtle
http://data.doremus.org/<group>/<uuid>[/<suffix>]
```

The `suffix` is for properties that normally are unique for a certain class

Examples:

``` turtle
http://data.doremus.org/expression/10430039-5e02-3ec8-bc6b-b433cdde13c7

http://data.doremus.org/expression/10430039-5e02-3ec8-bc6b-b433cdde13c7/dedication
```

This should be a mapping table (not complete) that covers classes we convert until now.

| Class | group |
| --- | --- |
| F22_SelfContainedExpression | expression |
| F28_ExpressionCreation | event |
| F14_IndividualWork | work |
| F15_ComplexWork | work |
| F42_RepresentativeExpressionAssignment | event |
| F24_PublicationExpression | publication ? |
| F30_PublicationEvent | event |
| F19_PublicationWork | work ? |
| F31_Performance | performance |
| F25_PerformancePlan | expression |
| F40_IdentifierAssignment | event |
| F50_ControlledAccessPoint |  |

**Comment by pierrechoffe** [comment]
The question is : who should this simplification help and how ?

**Comment by pasqLisena** [comment]
@pierrechoffe good question.

Some short answers:
- the URIs are shorter;
- the URIs are more easily readable by humans;
- we remove the redundancy Self_Contained_Expression/F22/ (2 levels, only one concept);
- less different values after the first `/` means less rule on the server (and less errors).

I am going to go deeper in [this](http://philarcher.org/diary/2013/uripersistence/)

**Comment by rtroncy** [comment]
@pierrechoffe This follows the best practices established by numerous agencies (thanks @pasqLisena for mentioning Phil Archer's work) and a good balance between completely opaque URI and fully transparent URI. This sticks better with the REST principles too, so when we will enable an ELDA configuration on top of the data, we could GET the collection ```<http://data.doremus.org/work>``` to get all resources that are F14 or F15.

**Comment by rtroncy** [comment]
I have updated the table and I would suggest to drop F40 and F50 for now

**Comment by pierrechoffe** [comment]
@pasqLisena @rtroncy 

>we could GET the collection <http://data.doremus.org/work> to get all resources that are F14 or F15.

But since all expressions are under one global "expression", how would you specifically look for F22 expressions?
Or, to put it differently, how would you diferentiate a notated music expression (Beethoven's Moonlight Sonata) from a performed music expression (Daniel Barenboim's rendering of this sonata)?
NB: we have changes coming in Performances, which now *create performed expressions* M43.

**Comment by rtroncy** [comment]
I cannot comment what I don't see yet in the document. Your comment is about expressions citing something about works which is confusion as well. I have suggested that F31 get a /performance top-level object in the URI. To answer your question, there is no problem to get anything as specific as you want: resources have type, property values, on which you can filter as much as you want. I think you get confused once more: a URI is an identifier, full stop. Embedding semantics in an identifier is a dangerous slope. You don't filter things based on identifiers, but based on property values attached to resources. Why we have decided to go for semi-transparent URI (as opposed to fully opaque ones) is for human readibility and ease of re-use but obviously, no real application will filter data based on a URI pattern!

**Comment by pasqLisena** [comment]
Should we add a 'resource' before the type?
I mean like `http://data.doremus.org/resource/expression/10430039-5e02-3ec8-bc6b-b433cdde13c7`

See [Phil Archer's work](http://philarcher.org/diary/2013/uripersistence/), par 3.1.1

**Comment by rtroncy** [comment]
tl;dr: NO

Slightly longer version: Phil's web site seems to be down this morning, but please, let's not start a [httprange-14](https://en.wikipedia.org/wiki/HTTPRange-14) discussion :-) In DOREMUS, it's clear we are in the metadata world and I don't think distinguishing the real-world object (concrete entity) from its description is useful. Similarly, we will probably not follow the guidelines described in [httpRange-14, Cool URIs & FRBR](http://efoundations.typepad.com/efoundations/2009/02/httprange14-cool-uris-frbr.html).

**Comment by pasqLisena** [comment]
We put `F42_RepresentativeExpressionAssignment` in the `expression` collection, but it is not a derivative of expression, but of event.
Should we change it? @rtroncy

**Comment by pierrechoffe** [comment]
@pasqLisena @rtroncy 
>We put F42_RepresentativeExpressionAssignment in the expression collection, but it is not a derivative of expression, but of event.

Absolutely, F42 is an event.
F40 is also an event
F50 is an identifier.
Note also that F24 is an expression. If you want to create a category "publication", then F30 and F19 would probably be part of "publication", right?
Actually, the question is : what kind of categories would you like to create? Work, Expression, Event are clearly CRM/FRBRoo oriented. Performance is an event, should it be part of the Event category or should it be a category in itself (which makes sense btw)?

**Comment by pasqLisena** [comment]
> Absolutely, F42 is an event.

Corrected.

About the other questions, I think that the sense is creating logical macro-groups for easy retrieving (basically through ELDA, as @rtroncy commented [some time ago](#issuecomment-242505598)).


**BTW** I am going to commit these modification. If we want to change something after, on the coding-side is a trivial operation.

**Comment by pasqLisena** [comment]
First export with the new policy:
https://drive.google.com/file/d/0ByLy_xr6iMAYdEhVT2Y0Q0tOOUE/view?usp=sharing

**Comment by rtroncy** [comment]
Thanks @pasqLisena for the implementation. Indeed F42 and F40 are events. I'm ignoring F50 on purpose for now. This needs to be re-designed anyway.

Regarding F24, F30 and F19 ... it really depends if you consider 'publication' to be the process itself or the result of the process. I suggest to keep it like this for the moment, but I know that it will become clearer when we will implement the conversion of the metadata regarding publications ... meaning, when we will have the mapping rules :-) For this reason, I will-reopen this issue and add a few question marks in the table in the original issue description.

**Comment by rtroncy** [comment]
See also https://github.com/DOREMUS-ANR/marc2rdf/blob/master/URI.patterns.md
